## 데이터 준비


In [2]:
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

In [3]:
digit = load_digits(as_frame=True)

In [4]:
digit.data.head()

pixel_0_0  pixel_0_1  pixel_0_2  pixel_0_3  pixel_0_4  pixel_0_5  \
0        0.0        0.0        5.0       13.0        9.0        1.0   
1        0.0        0.0        0.0       12.0       13.0        5.0   
2        0.0        0.0        0.0        4.0       15.0       12.0   
3        0.0        0.0        7.0       15.0       13.0        1.0   
4        0.0        0.0        0.0        1.0       11.0        0.0   

   pixel_0_6  pixel_0_7  pixel_1_0  pixel_1_1  ...  pixel_6_6  pixel_6_7  \
0        0.0        0.0        0.0        0.0  ...        0.0        0.0   
1        0.0        0.0        0.0        0.0  ...        0.0        0.0   
2        0.0        0.0        0.0        0.0  ...        5.0        0.0   
3        0.0        0.0        0.0        8.0  ...        9.0        0.0   
4        0.0        0.0        0.0        0.0  ...        0.0        0.0   

   pixel_7_0  pixel_7_1  pixel_7_2  pixel_7_3  pixel_7_4  pixel_7_5  \
0        0.0        0.0        6.0       13.0       10.0        0.0   
1        0.0        0.0        0.0       11.0       16.0       10.0   
2        0.0        0.0        0.0        3.0       11.0       16.0   
3        0.0        0.0        7.0       13.0       13.0        9.0   
4        0.0        0.0        0.0        2.0       16.0        4.0   

   pixel_7_6  pixel_7_7  
0        0.0        0.0  
1        0.0        0.0  
2        9.0        0.0  
3        0.0        0.0  
4        0.0        0.0  

[5 rows x 64 columns]

In [5]:
digit.target

0       0
1       1
2       2
3       3
4       4
       ..
1792    9
1793    0
1794    8
1795    9
1796    8
Name: target, Length: 1797, dtype: int32

In [6]:
digit.target_names

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [10]:
digit.target.value_counts()

target
3    183
1    182
5    182
4    181
6    181
9    180
7    179
0    178
2    177
8    174
Name: count, dtype: int64

In [7]:
digit.data.describe()

pixel_0_0    pixel_0_1    pixel_0_2    pixel_0_3    pixel_0_4  \
count     1797.0  1797.000000  1797.000000  1797.000000  1797.000000   
mean         0.0     0.303840     5.204786    11.835838    11.848080   
std          0.0     0.907192     4.754826     4.248842     4.287388   
min          0.0     0.000000     0.000000     0.000000     0.000000   
25%          0.0     0.000000     1.000000    10.000000    10.000000   
50%          0.0     0.000000     4.000000    13.000000    13.000000   
75%          0.0     0.000000     9.000000    15.000000    15.000000   
max          0.0     8.000000    16.000000    16.000000    16.000000   

         pixel_0_5    pixel_0_6    pixel_0_7    pixel_1_0    pixel_1_1  ...  \
count  1797.000000  1797.000000  1797.000000  1797.000000  1797.000000  ...   
mean      5.781859     1.362270     0.129661     0.005565     1.993879  ...   
std       5.666418     3.325775     1.037383     0.094222     3.196160  ...   
min       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
25%       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
50%       4.000000     0.000000     0.000000     0.000000     0.000000  ...   
75%      11.000000     0.000000     0.000000     0.000000     3.000000  ...   
max      16.000000    16.000000    15.000000     2.000000    16.000000  ...   

         pixel_6_6    pixel_6_7    pixel_7_0    pixel_7_1    pixel_7_2  \
count  1797.000000  1797.000000  1797.000000  1797.000000  1797.000000   
mean      3.725097     0.206455     0.000556     0.279354     5.557596   
std       4.919406     0.984401     0.023590     0.934302     5.103019   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     1.000000   
50%       1.000000     0.000000     0.000000     0.000000     4.000000   
75%       7.000000     0.000000     0.000000     0.000000    10.000000   
max      16.000000    13.000000     1.000000     9.000000    16.000000   

         pixel_7_3    pixel_7_4    pixel_7_5    pixel_7_6    pixel_7_7  
count  1797.000000  1797.000000  1797.000000  1797.000000  1797.000000  
mean     12.089037    11.809126     6.764051     2.067891     0.364496  
std       4.374694     4.933947     5.900623     4.090548     1.860122  
min       0.000000     0.000000     0.000000     0.000000     0.000000  
25%      11.000000    10.000000     0.000000     0.000000     0.000000  
50%      13.000000    14.000000     6.000000     0.000000     0.000000  
75%      16.000000    16.000000    12.000000     2.000000     0.000000  
max      16.000000    16.000000    16.000000    16.000000    16.000000  

[8 rows x 64 columns]

In [18]:
X_train, X_test, y_train, y_test = train_test_split(
    digit.data, digit.target, random_state=42
)

In [19]:
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((1347, 64), (450, 64), (1347,), (450,))

## 모델 학습
타겟의 분포가 비슷하니 정확도를 평가지표로 삼아도 괜찮을 것이라 판단했습니다  

모든 값이 0~255의 범위 안에 있어 스케일링은 안해도 괜찮다 생각했습니다

In [20]:
from sklearn.tree import DecisionTreeClassifier

model_tree = DecisionTreeClassifier(random_state=42)
model_tree.fit(X_train, y_train)
y_pred_tree = model_tree.predict(X_test)

print(classification_report(y_test, y_pred_tree))

              precision    recall  f1-score   support

           0       0.93      0.91      0.92        43
           1       0.84      0.86      0.85        37
           2       0.89      0.84      0.86        38
           3       0.79      0.83      0.81        46
           4       0.91      0.89      0.90        55
           5       0.89      0.81      0.85        59
           6       0.93      0.96      0.95        45
           7       0.88      0.88      0.88        41
           8       0.69      0.71      0.70        38
           9       0.81      0.88      0.84        48

    accuracy                           0.86       450
   macro avg       0.86      0.86      0.86       450
weighted avg       0.86      0.86      0.86       450



In [25]:
from sklearn.ensemble import RandomForestClassifier

model_rfc = RandomForestClassifier(random_state=42)
model_rfc.fit(X_train, y_train)
y_pred_rfc = model_rfc.predict(X_test)

print(classification_report(y_test, y_pred_rfc))

              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       0.97      1.00      0.99        37
           2       1.00      1.00      1.00        38
           3       0.96      0.93      0.95        46
           4       0.98      1.00      0.99        55
           5       0.94      0.98      0.96        59
           6       0.98      0.98      0.98        45
           7       0.98      0.98      0.98        41
           8       0.95      0.95      0.95        38
           9       0.98      0.92      0.95        48

    accuracy                           0.97       450
   macro avg       0.97      0.97      0.97       450
weighted avg       0.97      0.97      0.97       450



In [26]:
from sklearn.svm import SVC

model_svm = SVC(random_state=42)
model_svm.fit(X_train, y_train)
y_pred_svm = model_svm.predict(X_test)

print(classification_report(y_test, y_pred_svm))

              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       0.97      1.00      0.99        37
           2       1.00      1.00      1.00        38
           3       0.96      0.93      0.95        46
           4       0.98      1.00      0.99        55
           5       0.94      0.98      0.96        59
           6       0.98      0.98      0.98        45
           7       0.98      0.98      0.98        41
           8       0.95      0.95      0.95        38
           9       0.98      0.92      0.95        48

    accuracy                           0.97       450
   macro avg       0.97      0.97      0.97       450
weighted avg       0.97      0.97      0.97       450



In [27]:
from sklearn.linear_model import SGDClassifier

model_sgd = SGDClassifier(random_state=42)
model_sgd.fit(X_train, y_train)
y_pred_sgd = model_sgd.predict(X_test)

print(classification_report(y_test, y_pred_sgd))

              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       0.97      0.92      0.94        37
           2       0.93      1.00      0.96        38
           3       1.00      0.93      0.97        46
           4       0.98      0.98      0.98        55
           5       0.98      0.92      0.95        59
           6       0.96      0.98      0.97        45
           7       0.98      0.98      0.98        41
           8       0.80      0.95      0.87        38
           9       0.94      0.92      0.93        48

    accuracy                           0.95       450
   macro avg       0.95      0.95      0.95       450
weighted avg       0.96      0.95      0.95       450



In [28]:
from sklearn.linear_model import LogisticRegression

model_logi = LogisticRegression(random_state=42)
model_logi.fit(X_train, y_train)
y_pred_logi = model_logi.predict(X_test)

print(classification_report(y_test, y_pred_logi))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        43
           1       0.97      0.97      0.97        37
           2       0.97      1.00      0.99        38
           3       0.98      0.96      0.97        46
           4       1.00      0.96      0.98        55
           5       0.95      0.97      0.96        59
           6       0.96      0.98      0.97        45
           7       1.00      0.98      0.99        41
           8       0.95      0.97      0.96        38
           9       0.96      0.96      0.96        48

    accuracy                           0.97       450
   macro avg       0.97      0.97      0.97       450
weighted avg       0.97      0.97      0.97       450



c:\Users\zzoon\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [33]:
from xgboost import XGBClassifier

model_xgb = XGBClassifier(random_state=42)
model_xgb.fit(X_train, y_train)
y_pred_xgb = model_xgb.predict(X_test)

print(classification_report(y_test, y_pred_xgb))

              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       0.97      1.00      0.99        37
           2       0.95      1.00      0.97        38
           3       1.00      0.93      0.97        46
           4       0.98      0.98      0.98        55
           5       0.97      0.98      0.97        59
           6       0.98      0.98      0.98        45
           7       0.98      0.98      0.98        41
           8       0.95      0.97      0.96        38
           9       0.98      0.96      0.97        48

    accuracy                           0.98       450
   macro avg       0.98      0.98      0.98       450
weighted avg       0.98      0.98      0.98       450



## XGB 파인튜닝

In [36]:
from sklearn.model_selection import RandomizedSearchCV
import numpy as np

In [98]:
from xgboost import XGBClassifier

params = {
    "n_estimators": list(range(100, 500)),
    "learning_rate": list(np.logspace(np.log10(0.0001), np.log10(0.1))),
    "max_depth": list(np.arange(3, 10)),
    # "reg_alpha": list(np.linspace(0, 1)),
    # "reg_lambda": list(np.linspace(0, 1)),
    "gamma": list(np.logspace(np.log10(0.01), np.log10(20))),
    "colsample_bytree": list(np.linspace(0.6, 1, 10)),
    "subsample": list(np.linspace(0.5, 1, 10)),
}

model_xgb = XGBClassifier(random_state=42, n_jobs=-1)
tuner = RandomizedSearchCV(model_xgb, params, cv=3, n_iter=50)
tuner.fit(X_train, y_train)

RandomizedSearchCV(cv=3,
                   estimator=XGBClassifier(base_score=None, booster=None,
                                           callbacks=None,
                                           colsample_bylevel=None,
                                           colsample_bynode=None,
                                           colsample_bytree=None, device=None,
                                           early_stopping_rounds=None,
                                           enable_categorical=False,
                                           eval_metric=None, feature_types=None,
                                           gamma=None, grow_policy=None,
                                           importance_type=None,
                                           interaction_constraints=None,
                                           learning_rate...
                                                          0.005963623316594642, ...],
                                        'max_depth': [3, 4, 5, 6, 7, 8, 9],
                                        'n_estimators': [100, 101, 102, 103,
                                                         104, 105, 106, 107,
                                                         108, 109, 110, 111,
                                                         112, 113, 114, 115,
                                                         116, 117, 118, 119,
                                                         120, 121, 122, 123,
                                                         124, 125, 126, 127,
                                                         128, 129, ...],
                                        'subsample': [0.5, 0.5555555555555556,
                                                      0.6111111111111112,
                                                      0.6666666666666666,
                                                      0.7222222222222222,
                                                      0.7777777777777778,
                                                      0.8333333333333333,
                                                      0.8888888888888888,
                                                      0.9444444444444444,
                                                      1.0]})

In [99]:
tuner.best_params_

{'subsample': 0.7777777777777778,
 'n_estimators': 494,
 'max_depth': 9,
 'learning_rate': 0.028117686979742307,
 'gamma': 0.040393513624019946,
 'colsample_bytree': 0.8666666666666667}

In [105]:
model_fin = XGBClassifier(
    random_state=42,
    subsample=0.7777777777777778,
    n_estimators=494,
    max_depth=9,
    learning_rate=0.028117686979742307,
    gamma=0.040393513624019946,
    colsample_bytree=0.8666666666666667,
)
model_fin.fit(X_train, y_train)
y_pred_fin = model_fin.predict(X_test)

print(classification_report(y_test, y_pred_fin))

              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       0.95      1.00      0.97        37
           2       1.00      1.00      1.00        38
           3       1.00      0.93      0.97        46
           4       0.98      0.98      0.98        55
           5       0.95      0.98      0.97        59
           6       0.98      0.98      0.98        45
           7       0.98      1.00      0.99        41
           8       0.92      0.95      0.94        38
           9       1.00      0.96      0.98        48

    accuracy                           0.98       450
   macro avg       0.98      0.98      0.98       450
weighted avg       0.98      0.98      0.98       450



## 결론

정확도 98%로 xgboost가 제일 성능이 좋아 랜덤서치를 해보았는데 기본 파라미터값이랑 성능차이가 거의 비슷하였습니다
 
